<launch>

  <!-- defines name of the namespace of the drone -->
  <arg name="UAV_NAME" default="$(env UAV_NAME)" />
  <arg name="RUN_TYPE" default="$(env RUN_TYPE)" />

  <arg name="standalone" default="true" />

  <!-- allows disabling the gui from command line -->
  <arg name="gui" default="true" />

  <!-- will it run using GNU debugger? -->
  <arg name="debug" default="false" />
  <arg unless="$(arg debug)" name="launch_prefix_debug" value=""/>
  <arg     if="$(arg debug)" name="launch_prefix_debug" value="debug_roslaunch"/>

  <!-- will it run as standalone nodelet or using a nodelet manager? -->
  <arg name="manager" default="$(arg UAV_NAME)_occupancy_grid_merger_manager" />
  <arg name="n_threads" default="8" />
  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg manager)"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>
  <arg name="map_frame" default="$(arg UAV_NAME)/hector_map" /> <!-- Frame for hector slam -->
  <arg name="map_frame_uav47" default="/uav47/hector_map" /> <!-- Frame for hector slam -->
  <arg name="hector_frame" default="$(arg UAV_NAME)/hector" /> <!-- Frame for hector slam -->
  <arg name="hector_frame_uav47" default="/uav47/hector" /> <!-- Frame for hector slam -->


  <!-- Namespace -->
  <group ns="$(arg UAV_NAME)">
      <!-- Transform for hector slam -->
    <node pkg="tf2_ros" type="static_transform_publisher" name="hector_origin_to_hector_map" args="0 0 0 0 0 0 1 $(arg UAV_NAME)/hector_origin $(arg map_frame)" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="hector_origin_to_hector_map_uav47" args="0 0 0 0 0 0 1 /uav47/hector_origin $(arg map_frame_uav47)" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="hector_map_to_hector" args="0 0 0 0 0 0 1 $(arg map_frame) $(arg hector_frame)" />
    <node pkg="tf2_ros" type="static_transform_publisher" name="hector_map_to_hector_uav47" args="0 0 0 0 0 0 1 $(arg map_frame_uav47) $(arg hector_frame_uav47)" />
    
    <node pkg="tf2_ros" type="static_transform_publisher" name="received_map_to_current_map" args="0 0 0 0 0 0 1 $(arg UAV_NAME)/fcu /uav47/fcu" />
      
      <!-- OccupancyGridMerger nodelet -->
    <node pkg="nodelet" type="nodelet" name="occupancy_grid_merger" args="$(arg nodelet) occupancy_grid_merger/OccupancyGridMerger $(arg nodelet_manager)" launch-prefix="$(arg launch_prefix_debug)" output="screen">

      <!-- Parameters loaded from launch file -->
      <param name="gui" type="bool" value="$(arg gui)"/>
      
      <param name="simulation" value="$(eval arg('RUN_TYPE') == 'simulation')"/>
      
      <param name="UAV_NAME" type="string" value="$(arg UAV_NAME)"/>

      <!-- ROS parameters config file -->
      <rosparam file="$(find occupancy_grid_merger)/config/default.yaml" />
      <rosparam file="$(find occupancy_grid_merger)/config/$(arg RUN_TYPE).yaml" />

      <!-- Subscribers -->
      <remap from="~rec_map1_in" to="hector_mapping/received_map1" /> 
      <remap from="~occupancy_grid_map_in" to="hector_mapping/map" /> 
      <remap from="~odom_uav_in" to="odometry/odom_main" />
      <remap from="~odom_rec1_in" to="odometry/odom_main_rec1" />
      <remap from="~control_manager_diagnostics_in" to="control_manager/diagnostics" />
      <remap from="~odom_gt_in" to="ground_truth" />
  

      <!-- Publishers -->
      <remap from="~og_map_new_out" to="~new_map" />
      <remap from="~og_map_merged_out" to="~merged_map" />
      <remap from="~reference_out" to="control_manager/reference" />
      <remap from="~occupied_cells_rec_out" to="~/visualization/occupied_cells_rec" /> <!-- visualization_marker -->
      <remap from="~occupied_cells_cur_out" to="~/visualization/occupied_cells_cur" /> <!-- visualization_marker -->
      <remap from="~consistent_cells_out" to="~/visualization/consistent_cells" /> <!-- visualization_marker -->
      <remap from="~generation_poses_out" to="~/visualization/generation_poses" />
      <remap from="~relative_map_pose_out" to="~relative_map_pose" /> 
      <remap from="~relative_pose_out" to="~relative_pose" /> 
      <remap from="~relative_pose_glob_map_out" to="~relative_glob_pose" /> 

      <!-- Service servers -->
      <remap from="~test_in" to="~test" />
      
    </node>

  </group>

</launch>
